### PR TITLE
Protocol folder manager class

### PR DESCRIPTION
**Pull request by ogeesan**
> As a developer, I want a protocol finding tool that can return the filepath of a desired protocol.

### CODE REVIEW

**Comment by austin2118ace** [comment]
cursory glance looks good. can you add a `list_protocols` function that outputs all the protocols it found by name.

File: bpod_rig/protocols/manager.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+        Examples
+        --------
+            >>> manager = ProtocolManager(Path("/protocols"))
+            >>> manager.find_protocol_file("MyProtocol")
+            Path("/protocols/MyProtocol/MyProtocol.py")
+            >>> manager.find_protocol_file("subfolder/MyProtocol")
+            Path("/protocols/subfolder/MyProtocol/MyProtocol.py")
+        """
+        # Normalize path separators to use forward slashes
+        protocol_name_normalized = protocol_name.replace("\\", "/")
```
Comment (austin2118ace) [inline bpod_rig/protocols/manager.py:61]: thanks microslop

File: bpod_rig/protocols/manager.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        # Split into path components
+        protocol_basename = protocol_name_normalized.split("/")[-1]
+        matches = []
+        for protocol_file in self.protocols:
```
Comment (austin2118ace) [inline bpod_rig/protocols/manager.py:66]: would it make sense to use a dictionary instead of the set so you don't have to do this for-loop search?
Would the list of protocols ever become so large as to make a difference?

File: bpod_rig/protocols/manager.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        for protocol_file in self.protocol_dir.glob("**/*/*.py"):
+            if protocol_file.stem == protocol_file.parent.name:
+                self.protocols.add(protocol_file.absolute())
+
+    def find_protocol_file(self, protocol_name: str) -> Path:
+        r"""Find a protocol file in the protocol directory.
+
+        Searches for a protocol file matching the given name. The protocol file
+        should be located in a directory structure like:
+            protocol_dir/[subfolder/]protocol_name/protocol_name.py
```
Comment (austin2118ace) [inline bpod_rig/protocols/manager.py]: square bracket swapped with slash

File: bpod_rig/protocols/manager.py [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        if len(matches) == 0:
+            raise FileNotFoundError(
+                f"Protocol '{protocol_name}' not found in {self.protocol_dir}"
+            )
+        if len(matches) > 1:
+            match_paths = "\n  ".join(str(m) for m in matches)
+            raise ValueError(
+                f"Multiple protocols found matching '{protocol_name}':\n"
+                f"      {match_paths}\n"
+                f"Please specify more of the path (e.g., 'subfolder/{protocol_name}')"
```
Comment (austin2118ace) [inline bpod_rig/protocols/manager.py]: wording jumps as weird here, maybe `Please provide a more specific path` or `Please provide a more detailed path`

File: tests/bpod_rig/protocols/test_manager.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+        "def Protocol_unique1(bpod_session):\n    pass\n"
+    )
+
+    # Create subfolderA protocols
+    (protocol_folder / "subfolderA" / "Protocol_matching1").mkdir(
+        parents=True, exist_ok=True
+    )
+    (
+        protocol_folder / "subfolderA" / "Protocol_matching1" / "Protocol_matching1.py"
+    ).write_text("def Protocol_matching1(bpod_session):\n    pass\n")
```
Comment (austin2118ace) [inline tests/bpod_rig/protocols/test_manager.py]: maybe make a little generator function for all of these?

``` python
def generate_test_directory(*path, protocol_name):
    path = protocol_folder.joinpath(path)
    path.mkdir(parents=True, exist_ok=True)

    pyfile = path.joinpath(f"{protocol_name}.py"
    pyfile.write_text(f"def {protocol_name}(bpod_session):\n    pass \n")
```

File: tests/bpod_rig/protocols/test_manager.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    yield root_path / "protocols"
+    # Cleanup happens automatically with tmp_path_factory
+
+
+class TestFindProtocolFile:
+    """Test suite for finding protocol files."""
+
+    def test_basic_find(self, protocol_folder):
+        """Test finding a unique protocol at the root level."""
+        expected_path = protocol_folder / "Protocol_unique1" / "Protocol_unique1.py"
```
Comment (austin2118ace) [inline tests/bpod_rig/protocols/test_manager.py:80]: all these strings are used a lot, so maybe making a big list of constants at the beginning could make it easier to add future test cases?

**Review by austin2118ace — APPROVED** [review]
Solid work like always. I think this will provide a solid base for managing and searching for protocols

File: bpod_rig/protocols/manager.py (hunk 2 above)
Comment (ogeesan) [reply bpod_rig/protocols/manager.py:66]: No, in this situation we're using the `Path.match` method with a wildcard, because we may have `Protocols/ProtocolA/ProtocolA.py` and `Protocols/subfolder/ProtocolA/ProtocolA.py` and we need to see if we can find both.

File: bpod_rig/protocols/manager.py (hunk 4 above)
Comment (ogeesan) [reply bpod_rig/protocols/manager.py]: Good catch, I've cleaned up the wording.

File: bpod_rig/protocols/manager.py (hunk 3 above)
Comment (ogeesan) [reply bpod_rig/protocols/manager.py]: Old habits die hard

File: tests/bpod_rig/protocols/test_manager.py (hunk 5 above)
Comment (ogeesan) [reply tests/bpod_rig/protocols/test_manager.py]: Yeah good idea. I think I originally didn't do this because I thought that creating things more explicitly would be better, but if there are edge-cases we can add those more explicitly.

File: tests/bpod_rig/protocols/test_manager.py (hunk 6 above)
Comment (ogeesan) [reply tests/bpod_rig/protocols/test_manager.py:80]: I think this is where I would say that being explicit is more valuable. In `create_test_protocols_folder` the folder structure is described, and I think a test is easier to understand if we can follow the expected path explicitly like this.
